functional-test/src/test/groovy/org/rundeck/util/common/WaitUtils.groovy

```diff
@@ -0,0 +1,108 @@
+package org.rundeck.util.common
+
+import groovy.transform.TypeChecked
+import groovy.transform.stc.ClosureParams
+import groovy.transform.stc.FirstParam
+import groovy.transform.stc.FromString
+
+import java.time.Duration
+
+class WaitUtils  {
+
+    /** Waits for the resource to be in the expected state by retrieving it and evaluating its acceptance as many times as needed.
+     * @param retryableResourceRetriever executed multiple times to retrieve the current state of the R resource.
+     * @param resourceAcceptanceEvaluator evaluates the state of the resource and and returns true if the resource is accepted.
+     * @param timeout max duration to wait for the resource to reach the expected state.
+     * @param checkPeriod time to wait between each check.
+     * @return the resource that met the acceptance criteria.
+     * @throws ResourceAcceptanceTimeoutException if the timeout is reached waiting for the resource to reach the expected state.
+     */
+    @TypeChecked
+    static <R> R waitFor(
+            Closure<R> retryableResourceRetriever,
+            @ClosureParams(value = FromString, options = ["R"]) Closure<Boolean> resourceAcceptanceEvaluator = { R r -> !!r },
```

Comment:
- What does the `!!` operator exactly does? It's the second time i see it somewhere but can't find any info on it. Is groovy? Is java 11?
- This also sums to my concerns specifically about groovy closures, as they strength of being too flexible is also their weakness as you're never sure what they actually expect without IDE introspection. Maybe using `java.util.Function` would be better for these cases?

Reply:
- It has to do with groovy truthy and falsy. Consider: 
```
def foo = "bar"
println foo
println !foo
println !!foo

Output:
bar
false
true
```
In this case, returning an `r` violates typing, but `!!r` results in a `Boolean` (with compiler autoboxing)

Reply:
- Kinda hacky IMO. why don't just return true or let it throw an exception? 
- Using Function in practice should be mostly the same, but instead of declaring `@ClosureParams(value = FromString, options = ["R"]) Closure<Boolean> resourceAcceptanceEvaluator` you can just `Function<R, Boolean> resourceAcceptanceEvaluator`.